File: clouditor-engine-k8s/src/main/java/io/clouditor/discovery/k8s/KubernetesScanner.java
package io.clouditor.discovery.k8s;

import io.clouditor.credentials.KubernetesAccount;
import io.clouditor.discovery.MixInIgnore;
import io.clouditor.discovery.Scanner;
import io.clouditor.util.PersistenceManager;
import io.kubernetes.client.apis.CoreV1Api;
import java.io.IOException;
import java.util.function.Function;

public abstract class KubernetesScanner<T> extends Scanner<CoreV1Api, T> {

  public KubernetesScanner(Function<T, String> idGenerator, Function<T, String> nameGenerator) {
    super(null, idGenerator, nameGenerator);
  }

  @Override
  public void init() throws IOException {
    super.init();

    MAPPER.addMixIn(io.kubernetes.client.models.V1Probe.class, MixInIgnore.class);
    // TODO: provide a custom serializer for this instead of ignoring it
    MAPPER.addMixIn(io.kubernetes.client.custom.IntOrString.class, MixInIgnore.class);

    var account = PersistenceManager.getInstance().getById(KubernetesAccount.class, "Kubernetes");

    if (account == null) {
      throw new IOException("Kubernetes account not configured");
    }

    var builder = account.resolveCredentials();

    this.api = new CoreV1Api(builder.build());
  }
}

File: legacy/clouditor-engine-k8s/src/main/java/io/clouditor/discovery/k8s/KubernetesScanner.kt
package io.clouditor.discovery.k8s

import io.kubernetes.client.apis.CoreV1Api
import kotlin.Throws
import java.io.IOException
import io.kubernetes.client.models.V1Probe
import io.clouditor.discovery.MixInIgnore
import io.clouditor.credentials.KubernetesAccount
import io.clouditor.data_access_layer.HibernatePersistence
import io.clouditor.discovery.ScannerInfo
import io.clouditor.discovery.k8s.KubernetesScanner
import io.kubernetes.client.models.V1Pod
import io.clouditor.discovery.ScanException
import io.kubernetes.client.models.V1PodList
import io.kubernetes.client.ApiException
import io.kubernetes.client.models.NetworkingV1beta1Ingress
import io.kubernetes.client.apis.NetworkingV1beta1Api
import com.fasterxml.jackson.annotation.JsonTypeName
import io.clouditor.credentials.CloudAccount
import io.kubernetes.client.models.V1NamespaceList
import io.kubernetes.client.util.credentials.AccessTokenAuthentication
import com.auth0.jwt.interfaces.DecodedJWT
import com.auth0.jwt.interfaces.Claim
import io.clouditor.discovery.Scanner
import io.kubernetes.client.util.credentials.KubeconfigAuthentication
import io.kubernetes.client.ApiClient
import io.kubernetes.client.custom.IntOrString
import java.util.function.Function

abstract class KubernetesScanner<T>(idGenerator: Function<T, String>?, nameGenerator: Function<T, String>?) :
    Scanner<CoreV1Api?, T>(null, idGenerator, nameGenerator) {

    @Throws(IOException::class)
    override fun init() {
        super.init()

        MAPPER.addMixIn(V1Probe::class.java, MixInIgnore::class.java)
        // TODO: provide a custom serializer for this instead of ignoring it
        MAPPER.addMixIn(IntOrString::class.java, MixInIgnore::class.java)

        val account = HibernatePersistence()[KubernetesAccount::class.java, "Kubernetes"]
            .orElseThrow { IOException("Kubernetes account not configured") }
            ?: throw IOException("Kubernetes account not configured")

        val builder = account.resolveCredentials()

        api = CoreV1Api(builder.build())
    }
}